src/qengine/state/opencl.cpp

```diff
@@ -10,80 +10,90 @@
 // See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
 // for details.
 
+#include <memory>
+
 #include "oclengine.hpp"
 #include "qengine_opencl.hpp"
 
 namespace Qrack {
 
 #define CMPLX_NORM_LEN 5
 
-void QEngineOCL::InitOCL()
+void QEngineOCL::SetDevice(const int& dID)
 {
-    clObj = OCLEngine::Instance();
+    if ((dID >= clObj->GetDeviceCount()) || (dID < -1)) {
+        throw "Invalid OpenCL device selection";
+    } else {
+        deviceID = -1;
+    }
+    device_context = clObj->GetDeviceContextPtr(deviceID);
+    context = device_context->context;
+    queue = device_context->queue;
+}
 
-    queue = *(clObj->GetQueuePtr());
-    cl::Context context = *(clObj->GetContextPtr());
+void QEngineOCL::InitOCL(int devID)
+{
+    clObj = OCLEngine::Instance();
+    SetDevice(devID);
 
     // create buffers on device (allocate space on GPU)
-    stateBuffer = cl::Buffer(context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_WRITE, sizeof(complex) * maxQPower, stateVec);
+    stateBuffer = std::make_shared<cl::Buffer>(
+        context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_WRITE, sizeof(complex) * maxQPower, stateVec);
     cmplxBuffer = cl::Buffer(context, CL_MEM_READ_ONLY, sizeof(complex) * 5);
```

Comment:
Why don’t these need to be recreated when SetDevice to a new device which potentially involves a new context?

Reply:
See my other comment about SetDevice. They are, now.